README.md
# ramsey/base

[![Source Code][badge-source]][source]
[![Latest Version][badge-release]][packagist]
[![Software License][badge-license]][license]
[![PHP Version][badge-php]][php]
[![Build Status][badge-build]][build]
[![Coverage Status][badge-coverage]][coverage]
[![Total Downloads][badge-downloads]][downloads]

Arbitrary base conversion and encoding/decoding for all the bases that are belong to us.

This project adheres to a [Contributor Code of Conduct][conduct]. By
participating in this project and its community, you are expected to uphold this
code.


## Installation

The preferred method of installation is via [Composer][]. Run the following
command to install the package and add it as a requirement to your project's
`composer.json`:

```bash
composer require ramsey/base
```


## Documentation

Check out the [documentation website][documentation] for detailed information
and code examples.


## Contributing

Contributions are welcome! Please read [CONTRIBUTING][] for details.


## Copyright and License

The ramsey/base library is copyright © [Ben Ramsey](https://benramsey.com)
and licensed for use under the MIT License (MIT). Please see [LICENSE][] for
more information.


[conduct]: https://github.com/ramsey/base/blob/master/.github/CODE_OF_CONDUCT.md
[composer]: http://getcomposer.org/
[documentation]: https://ramsey.github.io/base/
[contributing]: https://github.com/ramsey/base/blob/master/.github/CONTRIBUTING.md

[badge-source]: http://img.shields.io/badge/source-ramsey/base-blue.svg?style=flat-square
[badge-release]: https://img.shields.io/packagist/v/ramsey/base.svg?style=flat-square&label=release
[badge-license]: https://img.shields.io/packagist/l/ramsey/base.svg?style=flat-square
[badge-php]: https://img.shields.io/packagist/php-v/ramsey/base.svg?style=flat-square
[badge-build]: https://img.shields.io/travis/ramsey/base/master.svg?style=flat-square
[badge-coverage]: https://img.shields.io/coveralls/github/ramsey/base/master.svg?style=flat-square
[badge-downloads]: https://img.shields.io/packagist/dt/ramsey/base.svg?style=flat-square&colorB=mediumvioletred

[source]: https://github.com/ramsey/base
[packagist]: https://packagist.org/packages/ramsey/base
[license]: https://github.com/ramsey/base/blob/master/LICENSE
[php]: https://php.net
[build]: https://travis-ci.org/ramsey/base
[coverage]: https://coveralls.io/r/ramsey/base?branch=master
[downloads]: https://packagist.org/packages/ramsey/base


CHANGELOG.md
# ramsey/base Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security


## [1.0.0] - YYYY-MM-DD

### Added

* Feature A description
* Feature B description
* Feature C description


[Unreleased]: https://github.com/ramsey/base/compare/1.0.0...HEAD
[1.0.0]: https://github.com/ramsey/base/commits/1.0.0


tests/TestCase.php
<?php

declare(strict_types=1);

namespace Ramsey\Base\Test;

use Mockery\Adapter\Phpunit\MockeryTestCase;

class TestCase extends MockeryTestCase
{
}


tests/SymbolTest.php
<?php

declare(strict_types=1);

namespace Ramsey\Base\Test;

use Error;
use Ramsey\Base\Symbol;

use const Ramsey\Base\BASE_1;
use const Ramsey\Base\BASE_10;
use const Ramsey\Base\BASE_16;
use const Ramsey\Base\BASE_2;
use const Ramsey\Base\BASE_32;
use const Ramsey\Base\BASE_32_CROCKFORD;
use const Ramsey\Base\BASE_32_GEOHASH;
use const Ramsey\Base\BASE_32_HEX;
use const Ramsey\Base\BASE_32_Z;
use const Ramsey\Base\BASE_35;
use const Ramsey\Base\BASE_36;
use const Ramsey\Base\BASE_62_X;
use const Ramsey\Base\BASE_64;
use const Ramsey\Base\BASE_64_URL;
use const Ramsey\Base\BASE_8;
use const Ramsey\Base\BASE_85_RFC;
use const Ramsey\Base\BASE_85_ZMQ;
use const Ramsey\Base\BASE_91;

class SymbolTest extends TestCase
{
    public function testTableConstant(): void
    {
        $this->assertSame(
            [
                1 => '1',
                2 => '01',
                8 => '01234567',
                10 => '0123456789',
                16 => '0123456789abcdef',
                32 => 'abcdefghijklmnopqrstuvwxyz234567',
                '32-crockford' => '0123456789abcdefghjkmnpqrstvwxyz',
                '32-geohash' => '0123456789bcdefghjkmnpqrstuvwxyz',
                '32-hex' => '0123456789abcdefghijklmnopqrstuv',
                '32-z' => 'ybndrfg8ejkmcpqxot1uwisza345h769',
                35 => '0123456789abcdefghijklmnpqrstuvwxyz',
                36 => '0123456789abcdefghijklmnopqrstuvwxyz',
                '62-x' => '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwyz$&*',
                64 => 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/',
                '64-url' => 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_',
                '85-rfc' => '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz!#$%&()*+-;<=>?@^_`{|}~',
                '85-zmq' => '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-:+=^!/*?&<>()[]{}@%$#',
                91 => 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!#$%&()*+,./:;<=>?@[]^_`{|}~"',
            ],
            Symbol::TABLE
        );
    }

    public function testRulesConstant(): void
    {
        $this->assertSame(
            [
                BASE_62_X => [
                    0 => ['$', '&', '*'],
                    1 => ['x1', 'x2', 'x3'],
                    2 => 64,
                ],
            ],
            Symbol::RULES
        );
    }

    public function testRulesKeysConstants(): void
    {
        $this->assertSame(0, Symbol::SUBSTITUTE);
        $this->assertSame(1, Symbol::ACTUAL);
        $this->assertSame(2, Symbol::LENGTH);
    }

    public function testInstantiationDenied(): void
    {
        $this->expectException(Error::class);
        $this->expectExceptionMessage(
            'Call to private Ramsey\Base\Symbol::__construct() from context '
            . '\'Ramsey\Base\Test\SymbolTest\''
        );

        new Symbol();
    }
}


tests/ConstantsTest.php
<?php

declare(strict_types=1);

namespace Ramsey\Base\Test;

class ConstantsTest extends TestCase
{
    /**
     * @param string $constantName
     * @param string|int $expectedValue
     * @dataProvider provideConstantsTestValues
     */
    public function testConstants(string $constantName, $expectedValue): void
    {
        $this->assertSame($expectedValue, constant($constantName));
    }

    public function provideConstantsTestValues(): array
    {
        return [
            ['Ramsey\Base\BASE_1', 1],
            ['Ramsey\Base\BASE_2', 2],
            ['Ramsey\Base\BASE_8', 8],
            ['Ramsey\Base\BASE_10', 10],
            ['Ramsey\Base\BASE_16', 16],
            ['Ramsey\Base\BASE_32', 32],
            ['Ramsey\Base\BASE_32_CROCKFORD', '32-crockford'],
            ['Ramsey\Base\BASE_32_GEOHASH', '32-geohash'],
            ['Ramsey\Base\BASE_32_HEX', '32-hex'],
            ['Ramsey\Base\BASE_32_Z', '32-z'],
            ['Ramsey\Base\BASE_35', 35],
            ['Ramsey\Base\BASE_36', 36],
            ['Ramsey\Base\BASE_62_X', '62-x'],
            ['Ramsey\Base\BASE_64', 64],
            ['Ramsey\Base\BASE_64_URL', '64-url'],
            ['Ramsey\Base\BASE_85_RFC', '85-rfc'],
            ['Ramsey\Base\BASE_85_ZMQ', '85-zmq'],
            ['Ramsey\Base\BASE_91', 91],
        ];
    }
}


src/Symbol.php
<?php

/**
 * This file is part of the ramsey/base library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Ben Ramsey <ben@benramsey.com>
 * @license http://opensource.org/licenses/MIT MIT
 */

declare(strict_types=1);

namespace Ramsey\Base;

final class Symbol
{
    public const TABLE = [
        BASE_1 => '1',
        BASE_2 => '01',
        BASE_8 => '01234567',
        BASE_10 => '0123456789',
        BASE_16 => '0123456789abcdef',
        BASE_32 => 'abcdefghijklmnopqrstuvwxyz234567',
        BASE_32_CROCKFORD => '0123456789abcdefghjkmnpqrstvwxyz',
        BASE_32_GEOHASH => '0123456789bcdefghjkmnpqrstuvwxyz',
        BASE_32_HEX => '0123456789abcdefghijklmnopqrstuv',
        BASE_32_Z => 'ybndrfg8ejkmcpqxot1uwisza345h769',
        BASE_35 => '0123456789abcdefghijklmnpqrstuvwxyz',
        BASE_36 => '0123456789abcdefghijklmnopqrstuvwxyz',
        BASE_62_X => '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwyz$&*',
        BASE_64 => 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/',
        BASE_64_URL => 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_',
        BASE_85_RFC => '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz!#$%&()*+-;<=>?@^_`{|}~',
        BASE_85_ZMQ => '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-:+=^!/*?&<>()[]{}@%$#',
        BASE_91 => 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!#$%&()*+,./:;<=>?@[]^_`{|}~"',
    ];

    public const SUBSTITUTE = 0;
    public const ACTUAL = 1;
    public const LENGTH = 2;

    public const RULES = [
        BASE_62_X => [
            self::SUBSTITUTE => ['$', '&', '*'],
            self::ACTUAL => ['x1', 'x2', 'x3'],
            self::LENGTH => 64,
        ],
    ];

    /**
     * Disallow instantiation of this class.
     *
     * @codeCoverageIgnore
     */
    private function __construct()
    {
    }
}


src/Converter.php
<?php

/**
 * This file is part of the ramsey/base library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Ben Ramsey <ben@benramsey.com>
 * @license http://opensource.org/licenses/MIT MIT
 */

declare(strict_types=1);

namespace Ramsey\Base;

use Brick\Math\BigInteger;
use RuntimeException;

final class Converter
{
    /**
     * @param string|int $number
     * @param string|int $fromBase
     * @param string|int $toBase
     *
     * @return string
     */
    public function convert($number, $fromBase, $toBase): string
    {
        if ((int) $fromBase < 36) {
            $number = strtolower((string) $number);
        }

        $this->assertCharactersInBase($number, $fromBase);

        if ($fromBase !== BASE_10) {
            $number = $this->baseXtoBase10($number, $fromBase);
        }

        if ($toBase === BASE_10) {
            return (string) $number;
        }

        return $this->base10toBaseX($number, $toBase);
    }

    private function baseXtoBase10($number, $fromBase)
    {
        if ($fromBase === BASE_1) {
            return strlen($number);
        }

        $base = BigInteger::of(Symbol::RULES[$fromBase][Symbol::LENGTH] ?? (int) $fromBase);

        $number = str_replace(
            Symbol::RULES[$fromBase][Symbol::ACTUAL] ?? '',
            Symbol::RULES[$fromBase][Symbol::SUBSTITUTE] ?? '',
            $number
        );

        $number = array_reverse(str_split($number));
        $baseAlphabet = str_split(Symbol::TABLE[$fromBase]);

        $return = BigInteger::zero();
        foreach ($number as $position => $symbol) {
            $value = BigInteger::of(array_search($symbol, $baseAlphabet));
            $return = $return->plus($value->multipliedBy($base->power($position)));
        }

        return (string) $return;
    }

    private function base10toBaseX($number, $toBase)
    {
        $base = Symbol::RULES[$toBase][Symbol::LENGTH] ?? (int) $toBase;
        $value = BigInteger::of($number);
        $return = '';

        do {
            $return = Symbol::TABLE[$toBase][$value->mod($base)->toInt()] . $return;

            if ($base === 1) {
                $value = $value->minus(1);
                continue;
            }

            $value = $value->quotient($base);
        } while ($value->isGreaterThan(0));

        $return = str_replace(
            Symbol::RULES[$toBase][Symbol::SUBSTITUTE] ?? '',
            Symbol::RULES[$toBase][Symbol::ACTUAL] ?? '',
            $return
        );

        return $return;
    }

    private function assertCharactersInBase(string $number, $fromBase): bool
    {
        $symbols = Symbol::TABLE[$fromBase];

        // Get rid of any substitute symbols.
        $symbols = str_replace(
            Symbol::RULES[$fromBase][Symbol::SUBSTITUTE] ?? '',
            '',
            $symbols
        );

        // Escape special characters.
        $symbols = str_replace(
            ['/', ']', '$', '-'],
            ['\/', '\]', '\$', '\-'],
            $symbols
        );

        $expression = '/^([' . $symbols . ']';

        // Add any actual symbols to the expression.
        foreach ((Symbol::RULES[$fromBase][Symbol::ACTUAL] ?? []) as $symbol) {
            $expression .= '|' . $symbol;
        }

        $expression .= ')*$/';

        if (!preg_match($expression, $number)) {
            throw new RuntimeException(sprintf(
                'Invalid characters found in base%s value `%s\'',
                $fromBase,
                $number
            ));
        }

        return true;
    }
}


src/constants.php
<?php

/**
 * This file is part of the ramsey/base library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Ben Ramsey <ben@benramsey.com>
 * @license http://opensource.org/licenses/MIT MIT
 */

declare(strict_types=1);

namespace Ramsey\Base;

// @codeCoverageIgnoreStart
const BASE_1 = 1;
const BASE_2 = 2;
const BASE_8 = 8;
const BASE_10 = 10;
const BASE_16 = 16;
const BASE_32 = 32;
const BASE_32_CROCKFORD = '32-crockford';
const BASE_32_GEOHASH = '32-geohash';
const BASE_32_HEX = '32-hex';
const BASE_32_Z = '32-z';
const BASE_35 = 35;
const BASE_36 = 36;
const BASE_62_X = '62-x';
const BASE_64 = 64;
const BASE_64_URL = '64-url';
const BASE_85_RFC = '85-rfc';
const BASE_85_ZMQ = '85-zmq';
const BASE_91 = 91;
// @codeCoverageIgnoreEnd
